### PR TITLE
CR 1183 CE Add Room Number

### PR DESCRIPTION
**Pull request by Paul-Joel**
# Motivation and Context
Allow users requesting a CE fulfilment to enter a room number in addition the selected address

# What has changed
Updates to CE fulfilments, to allow the user on the 'Confirm Address' screen, and/or the 'Confirm Name and Address' screen to specify a 'room number'. 
If added at either point the room number value is either appended to the end of the surname, or if the surname is too long is provided as the optional 'title' value in the RHSvc request fulfilment endpoint
Unit tests added to test for the journey additions, and to ensure values and links are not present in non-CE journeys.

No Cucumber update is required to merge

# How to test?
Request a UAC or paper form fulfilment, and enter a CE as the address. Ensure the 'Add room number' link appears on the address confirmation screen, can be entered, and then appears on the address confirmation screen. Repeat, but set the value in the 'Confirm Name and Address' screen. When set it should be possible to change the value on both screens.
Journey should complete as normal, with the room number appended in confirm address displays, and at the sent confirmation screens

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1183

# Screenshots (if appropriate):

### CODE REVIEW

File: app/requests_handlers.py [hunk 1]
```diff
@@ -613,11 +626,35 @@ async def post(self, request):
...
+
+                if room_number:
+                    if len(attributes['last_name']) < 48:
+                        last_name = attributes['last_name'] + ', ' + room_number
+                        title = None
+                    else:
+                        last_name = attributes['last_name']
+                        title = room_number
+                else:
+                    last_name = attributes['last_name']
```
Comment (rpcatling) [inline app/requests_handlers.py:641]: I've just checked the UI  manually and there doesn't appear to be any limit to the surname entered (I was able to type a surname 243 characters long for example).  When we try to post this we get a rejection at RHSvc and an ERROR log on RHUI, and a "Something went wrong" page.  Shouldn't we limit the surname to 60 characters at the UI to match the server ?  Perhaps using the 'charCheckLimit' that you had on the room name ? We might like to think about this for some other fields also.

**Review by rpcatling — CHANGES_REQUESTED** [review]
Various comments, most minor

File: app/utils.py [hunk 2]
```diff
@@ -425,16 +425,27 @@ async def request_fulfilment_sms(request, case_id, tel_no, fulfilment_code_array
...
+            fulfilment_json = {
+                'caseId': case_id,
+                'title': title,
+                'forename': first_name,
+                'surname': last_name,
+                'fulfilmentCodes': fulfilment_code_array,
+                'dateTime': datetime.now(timezone.utc).isoformat(),
+                'clientIP': View.single_client_ip(request)
+            }
+        else:
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/utils.py]: Simplified

File: app/requests_handlers.py [hunk 3]
```diff
@@ -613,11 +626,35 @@ async def post(self, request):
...
+                    title = None
+
+                try:
+                    if title:
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/requests_handlers.py]: Simplified

File: app/templates/common-confirm-address.html [hunk 4]
```diff
@@ -72,4 +82,4 @@
         })
     }}
 
-{% endblock %}
\ No newline at end of file
+{%- endblock -%}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/templates/common-confirm-address.html]: Added

File: app/templates/common-enter-room-number.html [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    {{
+        onsButton({
+            'text': _('Continue'),
+            'classes': 'u-mb-xl u-mt-l'
+        })
+    }}
+
+    <p><a href="{{ cancel_link }}">{{ _('Cancel and return to the previous page') }}</a></p>
+
+{%- endblock -%}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/templates/common-enter-room-number.html]: Added

File: tests/test_data/address_index/uprn_valid_ce.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        "addressType": "PAF",
+        "historical": true,
+        "epoch": ""
+    },
+    "status": {
+        "code": 200,
+        "message": "Ok"
+    },
+    "errors": []
+}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/address_index/uprn_valid_ce.json]: Added

File: tests/test_data/address_index/uprn_valid_spg.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        "addressType": "PAF",
+        "historical": true,
+        "epoch": ""
+    },
+    "status": {
+        "code": 200,
+        "message": "Ok"
+    },
+    "errors": []
+}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/address_index/uprn_valid_spg.json]: Added

File: app/requests_handlers.py [hunk 8]
```diff
@@ -613,11 +626,35 @@ async def post(self, request):
                     fulfilment_code_array.append(available_fulfilments[0]['fulfilmentCode'])
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/requests_handlers.py:629]: It's on my list to resolve - I'll probably sort it with CR-1144 as that is changes to the same code

File: app/requests_handlers.py [hunk 9]
```diff
@@ -613,11 +626,35 @@ async def post(self, request):
...
-                                                            attributes['case_id'],
-                                                            attributes['first_name'],
-                                                            attributes['last_name'],
-                                                            fulfilment_code_array)
+                    room_number = attributes['roomNumber']
+                except KeyError:
+                    room_number = None
+
+                if room_number:
+                    if len(attributes['last_name']) < 48:
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/requests_handlers.py]: Changed to a variable, and added a comment at the declaration

File: app/requests_handlers.py (hunk 1 above)
Comment (Paul-Joel) [reply app/requests_handlers.py:644]: Will get a ticket raised to do cover this throughout RHUI

**Review by rpcatling — APPROVED** [review]

File: app/requests_handlers.py [hunk 10]
```diff
@@ -612,12 +628,29 @@ async def post(self, request):
                 else:
                     fulfilment_code_array.append(available_fulfilments[0]['fulfilmentCode'])
 
+                try:
+                    room_number = attributes['roomNumber']
+                except KeyError:
+                    room_number = None
+
+                if room_number:
+                    if len(attributes['last_name']) < last_name_char_limit:
```
Comment (PeterBochel) [inline app/requests_handlers.py:637]: Just wondering about the case where the current last_name is just below the last_name limit (like last_name say 47 chars long and would be say limit+3 long once the room number is added).
Should this code not be comparing the length what last_name + room_number would be against the limit.
The if statement its self seems odd. It implies that it's ok to have a last_name which is longer than its limit :/

**Review by PeterBochel — CHANGES_REQUESTED** [review]

**Review by PeterBochel — APPROVED** [review]

**Pull request merged**